tools/nightly_checkout.py

```diff
@@ -0,0 +1,343 @@
+    $ conda activate my-env
+"""
+import os
+import re
+import sys
+import json
+import glob
+import time
+import shutil
+import logging
+import tempfile
+import functools
+import contextlib
+import subprocess
+from ast import literal_eval
+from argparse import ArgumentParser
+
+
+LOGGER = logging.getLogger("conda-pytorch")
+URL_FORMAT = "{base_url}/{platform}/{dist_name}.tar.bz2"
+SHA1_RE = re.compile("([0-9a-fA-F]{40})")
+SPECS_TO_INSTALL = ("pytorch", "mypy", "pytest", "hypothesis", "ipython", "sphinx")
+
+
+def init_logging(level=logging.INFO):
+    """Start up the logger"""
+    logging.basicConfig(level=level)
+
+
+def check_in_repo():
+    """Ensures that we are in the PyTorch repo."""
+    if not os.path.isfile("setup.py"):
+        return "Not in root-level PyTorch repo, no setup.py found"
+    with open("setup.py") as f:
+        s = f.read()
+    if "PyTorch" not in s:
+        return "Not in PyTorch repo, 'PyTorch' not found in setup.py"
+
+
+def check_branch(branch):
+    """Checks that the branch name can be checked out."""
+    # first make sure actual branch name was given
+    if branch is None:
+        return "Branch name to checkout must be supplied with '-b' option"
+    # next check that the local repo is clean
+    cmd = ["git", "status", "--untracked-files=no", "--porcelain"]
+    p = subprocess.run(cmd, capture_output=True, check=True, text=True)
+    if p.stdout.strip():
+        return "Need to have clean working tree to checkout!\n\n" + p.stdout
+    # next check that the branch name doesn't already exist
+    cmd = ["git", "show-ref", "--verify", "--quiet", "refs/heads/" + branch]
+    p = subprocess.run(cmd, capture_output=True, check=False)
+    if not p.returncode:
+        return f"Branch {branch!r} already exists"
+
+
+@contextlib.contextmanager
+def timer(logger, prefix):
+    """Timed conetxt manager"""
+    start_time = time.time()
+    yield
+    logger.info(f"{prefix} took {time.time() - start_time:.3f} [s]")
+
+
+def timed(prefix):
+    """Decorator for timing functions"""
+    def dec(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            LOGGER.info(prefix)
+            with timer(LOGGER, prefix):
+                return f(*args, **kwargs)
+        return wrapper
+    return dec
+
+
+@timed("Solving conda environment")
+def conda_solve(name=None, prefix=None):
+    """Performs the conda solve and splits the deps from the package."""
+    # compute what environment to use
+    if prefix is not None:
+        existing_env = True
+        env_opts = ["--prefix", prefix]
+    elif name is not None:
+        existing_env = True
+        env_opts = ["--name", name]
+    else:
+        # create new environment
+        existing_env = False
+        env_opts = ["--name", "pytorch-deps"]
+    # run solve
+    if existing_env:
+        cmd = ["conda", "install", "--yes", "--dry-run", "--json",
+               "-c", "pytorch-nightly", "-c", "conda-forge"]
```

Comment:
Ok! conda-forge and defaults are binary compatible, but if pytorch is specifically designed for defaults, then this should be defaults (or just not exist).  Not existing seems safest so that it relies on user settings